core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java

```diff
@@ -419,15 +384,64 @@ public void reportingEntryPublished(TestIdentifier testIdentifier, ReportEntry e
         }
     }
 
-    private void trimStackTrace(Class<?> testClass, Throwable throwable) {
-        if (testClass != null) {
+    private String getTestClassNameFromSource(Optional<TestSource> optionalTestSource) {
+        if (optionalTestSource.isPresent()) {
+            var testSource = optionalTestSource.get();
+            if (testSource instanceof MethodSource) {
+                return ((MethodSource) testSource).getJavaClass().getName();
+            } else if (testSource instanceof ClassSource) {
+                return ((ClassSource) testSource).getJavaClass().getName();
+            } else if (testSource.getClass().getName().equals("com.tngtech.archunit.junit.FieldSource")) {
+                try {
+                    return (String) testSource.getClass().getMethod("getClassName").invoke(testSource);
+                } catch (ReflectiveOperationException e) {
+                    throw new RuntimeException("Failed to read className reflectively from " + testSource, e);
+                }
+            }
+        }
+        return "";
+    }
+
+    private Optional<ClassLoader> getTestClassLoaderFromSource(Optional<TestSource> optionalTestSource) {
+        if (optionalTestSource.isPresent()) {
+            var testSource = optionalTestSource.get();
+            if (testSource instanceof MethodSource) {
+                return Optional.of(((MethodSource) testSource).getJavaClass().getClassLoader());
+            } else if (testSource instanceof ClassSource) {
+                return Optional.of(((ClassSource) testSource).getJavaClass().getClassLoader());
+            }
+            // no com.tngtech.archunit.junit.FieldSource here as it doesn't provide a Class object
+        }
+        return Optional.empty();
+    }
+
+    private String getDisplayNameFromIdentifier(TestIdentifier testIdentifier) {
+        if (testIdentifier.getSource().isPresent()) {
+            var testSource = testIdentifier.getSource().get();
+            if (testSource instanceof MethodSource) {
+                var methodSource = (MethodSource) testSource;
+                return methodSource.getJavaClass().getSimpleName() + "#" + testIdentifier.getDisplayName();
+            } else if (testSource.getClass().getName().equals("com.tngtech.archunit.junit.FieldSource")) {
+                try {
+                    return (String) testSource.getClass().getMethod("getFieldName").invoke(testSource)
```

Comment:
Just realized that this needs to get the class name, not the field name.

Reply:
I don't like it, but maybe it is ok for the short term. The only other option I can think of is to make this pluggable and create a quarkus-archunit Quarkiverse extension to support it. 

If this is all that is needed then maybe we can live with it, but if there are any other hacks required then this should be pulled into its own module IMHO.

Reply:
AFAICS, this is the only necessary "hack" for AU (actually only the `className` call, `fieldName` is not relevant and I will remove it).

For a quarkiverse extension we would need to denfine an SPI, right?

Alternatively, it would be nice if JUnit would offer either an optional method in `TestSource` returning the class name or a separate interface that a source can implement additionally.

Reply:
I removed that second reflective call (`fieldName`) and fixed a regression w.r.t. `testClassUsages.updateTestData(...)`.

I could improve that workaround if we decide to actually _load_ the class via the name that is provided by `com.tngtech.archunit.junit.FieldSource`. This way we could stick with `Class<?>` as before.
But I'm not sure this is a good idea.
Maybe AU can provide the actual class object via that `FieldSource`. We'd still need to get it via reflection, but we wouldn't need to juggle with a `String` class name.

Reply:
> Maybe AU can provide the actual class object via that `FieldSource`. We'd still need to get it via reflection, but we wouldn't need to juggle with a `String` class name.

https://github.com/TNG/ArchUnit/pull/800

Reply:
@geoand do you have anything to add to this?

Reply:
> > Maybe AU can provide the actual class object via that `FieldSource`. We'd still need to get it via reflection, but we wouldn't need to juggle with a `String` class name.
> 
> [TNG/ArchUnit#800](https://github.com/TNG/ArchUnit/pull/800)

This was just merged and will be part of 0.23 which will be released this weekend. I'll adjust this PR afterwards.
@stuartwdouglas Would be nice to agree (or not) on the (temporary) hackish workaround.

Reply:
I pushed an adjustment to ArchUnit 0.23.0 which was released today. It's still a reflection hack, but since we can now get the full class object, the change is less invasive than previously.
I double checked with my actual, more complex project (not just the small reproducer).

@stuartwdouglas please have a look. I won't merge before your "ok".

I'll also add the backport label (IIRC @gsmet was planning to release 2.7.2 very soon).
With that we would be pressing 2.7.2+ users to update to AU 0.23.0 (given they use continuous testing), which might be slighty debatable, but I think it's ok. I'll also add something to the migration guide.

Reply:
I think the hack is ok for now. If we end up needing more of this sort of thing then we should add a SPI.

Reply:
Turns out all those filters (e.g. `JunitTestRunner.RegexFilter`) don't cover that `FieldSource` yet (obviously), so there is no way to exclude a certain field-style ArchUnit test. So SPI it is... 😒 
Also, I'm seeing more strange classloading issues in some of my tests (which is the actual motivation for the exclusion attempt).
E.g. one fails right away not finding an interface (which is used in `org.mapstruct.Mapper.config()`) and other tests working with class objects are failing after a reload (caused by code changes). The latter might show a general issue, not only related to ArchUnit.
I'm also seeing an annotation class loading issue in a custom `org.junit.jupiter.api.extension.BeforeEachCallback` (only CT).

Time to create some issues and reproducers...